test/compose.js

```diff
@@ -0,0 +1,176 @@
+  let slug = `${COMPOSE_VERSION}-${platform}-${arch}`;
+  return { url, slug };
+}
+
+/**
+Docker compose related utilities...
+*/
+
+class Compose {
+  constructor(composeBin) {
+    let dockerConf = dockerOpts();
+
+    if (dockerConf.host) {
+      let parsedHost = URL.parse(dockerConf.host);
+      this.host = parsedHost.host;
+    } else {
+      this.host = 'localhost';
+    }
+
+    // Use remote host if given otherwise assume localhost...
+    this.bin = composeBin;
+    this.docker = new Docker(dockerConf);
+    debug('compose initialized', { bin: composeBin, host: this.host });
+  }
+
+  /**
+  Bring a docker-compose yaml stack online...
+  */
+  async up(cwd) {
+    return execDebug(`${this.bin} up -d --no-recreate`, { cwd });
+  }
+
+  async killAll(cwd) {
+    return execDebug(`${this.bin} kill`, { cwd });
+  }
+
+  /**
+  Start running a single instance of a given service.
+  */
+  async run(cwd, service) {
+    let [service] = await execDebug(`${this.bin} run -d --service-ports ${service}`, {
+      cwd
+    });
+
+    return service.trim();
+  }
+
+  async portByName(cwd, service, port) {
+    let [result] = await execDebug(`${this.bin} port ${service} ${port}`, {
+      cwd
+    });
+
+    let [host, port] = result.trim().split(':');
+    if (port) {
+      return parseInt(port, 10);
+    }
+    throw new Error(`Unknown port ${port} for service ${service}`);
+  }
+
+  async portById(containerId, port) {
+    let inspect = await this.inspect(containerId);
+    let ports = inspect.NetworkSettings.Ports;
+    let port = ports[`${port}/tcp`];
+
+    if (!port) {
+      throw new Error(`Unknown port ${port} for container ${containerId}`);
+    }
+    return parseInt(port[0].HostPort, 10);
+  }
+
+  /**
+  Ensure container is both killed and removed.
+  */
+  async destroy(containerId) {
+    let container = this.docker.getContainer(containerId);
+    await container.kill({});
+    await container.remove({});
+  }
+
+  async ps(cwd) {
+    let [stdout] = await execDebug(`${this.bin} ps -q`, {
+      cwd
+    });
+
+    return stdout.trim().split('\n').map((v) => {
+      return v.trim();
+    });
+  }
+
+  /**
+  Inspect a container given it's container id (note you can get this via ps)
+  */
+  async inspect(containerId) {
+    let container = this.docker.getContainer(containerId);
+    return await container.inspect();
+  }
+
+  async version() {
+    let [version] = await execDebug(`${this.bin} --version`);
+    let [, number] = version.trim().split(' ');
+    return number;
+  }
+}
+
+/**
+Download and install docker-compose.
+
+@param {String} [installPath] install path for compose binaries.
+*/
+export async function install(installPath = COMPOSE_INSTALL_ROOT) {
+  let { url, slug } = composeLocationConfig();
+  debug('compose install', { url, slug });
+
+  // Ensure compose install root exists...
+  if (!await fs.exists(installPath)) {
+    await fs.mkdir(installPath);
```

Comment:
There is a very little chance to get a race condition between fs.exists() and fs.mkdir(). I'd rather call fs.mkdir() and ignore EEXIST.

Also would be better to set mode to something safer than default 777.